File: Graph/number_of_islands.cc
// Link: https://leetcode.com/problems/number-of-islands/
// Approach: Perform BFS starting from a node of value 1. Mark all connected nodes as visited, so that you do not visit them again.
//           Increment the counter as you encounter unvisited 1 (in numIslands function).

#include <queue>

class Solution
{
    private:
        queue<pair<int, int>> node_queue;
        
        void bfs(int i, int j, vector<vector<char>>& grid)
        {
            pair<int, int> p;
            int x = -1, y = -1, m = grid.size(), n = grid[0].size();
            node_queue.push(make_pair(i, j));
            
            while(node_queue.size()>0)
            {
                p = node_queue.front();
                x = p.first;
                y = p.second;
                node_queue.pop();
                
                if(x>0)
                {
                    if(grid[x-1][y] == '1')
                    {
                        node_queue.push(make_pair(x-1, y));
                        grid[x-1][y] = '-';
                    }
                }
                
                if(x<(m-1))
                {
                    if(grid[x+1][y] == '1')
                    {
                        node_queue.push(make_pair(x+1, y));
                        grid[x+1][y] = '-';
                    }
                }
                
                if(y>0)
                {
                    if(grid[x][y-1] == '1')
                    {
                        node_queue.push(make_pair(x, y-1));
                        grid[x][y-1] = '-';
                    }
                }
                
                if(y<(n-1))
                {
                    if(grid[x][y+1] == '1')
                    {
                        node_queue.push(make_pair(x, y+1));
                        grid[x][y+1] = '-';
                    }
                }
            }
        }
    
    public:
        int numIslands(vector<vector<char>>& grid)
        {
            int m = grid.size();
            
            if(m == 0)
            {
                return 0;
            }
            
            int n = grid[0].size(), counter = 0;
            
            for(int i=0;i<m;i++)
            {
                for(int j=0;j<n;j++)
                {
                    if(grid[i][j] == '1')
                    {
                        counter++;
                        bfs(i, j, grid);
                    }
                }
            }
            return counter;
        }
};

File: Graph/number_of_islands.py
# Link: https://leetcode.com/problems/number-of-islands/

# Approach: Explore all the cells using for loops. Now at each cell when you encounter "1", append the island_counter and att that coordinates to the queue for BFS. Explore all 
# neighbouring land using BFS and mark each visited when adding to queue. Finally return island_counter.

from collections import deque

class Solution(object):
    def numIslands(self, grid):
        m = len(grid)
        
        if m == 0:
            return 0
        
        n, island_counter, queue = len(grid[0]), 0, deque()
        
        if n == 0:
            return 0
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    island_counter += 1
                    queue.append((i, j))
                    
                    # BFS
                    while queue:
                        y, x = queue.popleft()
                        
                        
                        if x-1 >= 0 and grid[y][x-1] == "1":
                            grid[y][x-1] = "-1"
                            queue.append((y, x-1))
                            
                        if x+1 < n and grid[y][x+1] == "1":
                            grid[y][x+1] = "-1"
                            queue.append((y, x+1))
                            
                        if y-1 >= 0 and grid[y-1][x] == "1":
                            grid[y-1][x] = "-1"
                            queue.append((y-1, x))
                        
                        if y+1 < m and grid[y+1][x] == "1":
                            grid[y+1][x] = "-1"
                            queue.append((y+1, x))
        return island_counter